File: app/src/main/java/com/soumikshah/investmenttracker/view/InvestmentAdapter.java
package com.soumikshah.investmenttracker.view;

import android.content.Context;
import android.text.Html;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.soumikshah.investmenttracker.R;
import com.soumikshah.investmenttracker.database.model.Investment;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import static android.view.View.GONE;

public class InvestmentAdapter extends RecyclerView.Adapter<InvestmentAdapter.MyViewHolder> {
    private Context context;
    private List<Investment> InvestmentsList;
    int mExpandedPosition = -1;
    public static class MyViewHolder extends RecyclerView.ViewHolder {
        public TextView investmentName;
        public TextView investmentAmount;
        public TextView investmentMedium;
        public TextView interestToBePaid;
        public TextView investmentCategory;
        public TextView investmentDate;
        public TextView investmentMonth;
        public RelativeLayout viewBackground, viewForeground;
        public LinearLayout otherDetails;


        public MyViewHolder(View view) {
            super(view);
            investmentName = view.findViewById(R.id.investment);
            investmentAmount = view.findViewById(R.id.investmentAmount);
            investmentMedium = view.findViewById(R.id.investmentMedium);
            investmentCategory = view.findViewById(R.id.investmentCategory);
            investmentDate = view.findViewById(R.id.investedDate);
            investmentMonth = view.findViewById(R.id.investedNumberOfMonths);
            interestToBePaid = view.findViewById(R.id.interestToBePaid);
            viewBackground = view.findViewById(R.id.view_background);
            viewForeground = view.findViewById(R.id.view_foreground);
            otherDetails = view.findViewById(R.id.otherDetails);
            //timestamp = view.findViewById(R.id.timestamp);
        }
    }

    public InvestmentAdapter(Context context, List<Investment> InvestmentsList) {
        this.context = context;
        this.InvestmentsList = InvestmentsList;
    }

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.investment_list_row, parent, false);

        return new MyViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(final MyViewHolder holder, final int position) {
        final Investment investment = InvestmentsList.get(position);

        holder.investmentName.setText(investment.getInvestmentName());
        holder.investmentAmount.setText(String.format("Rs.%s", String.valueOf(investment.getInvestmentAmount())));

        String investmentMedium = "Invested in: <b>"+investment.getInvestmentMedium()+"</b>";
        SimpleDateFormat sim = new SimpleDateFormat("dd/MM/YYYY",Locale.ENGLISH);
        String investmentCategory = "Investment Type: <b>"+investment.getInvestmentCategory()+"</b>";
        String investmentDate = "Invested on: <b>"+sim.format(investment.getInvestmentDate())+"</b>";
        String investmentMonth = "Invested for: <b>"+investment.getInvestmentMonth()+"</b>";
        String investmentInterestToBePaid = "Interest to be paid: <b>"+investment.getInvestmentPercent()+"</b>";

        holder.investmentMedium.setText(Html.fromHtml(investmentMedium));
        holder.investmentCategory.setText(Html.fromHtml(investmentCategory));
        holder.investmentDate.setText(Html.fromHtml(investmentDate));
        holder.investmentMonth.setText(Html.fromHtml(investmentMonth));
        holder.interestToBePaid.setText(Html.fromHtml(investmentInterestToBePaid));

        if(investment.getInvestmentPercent()==0.0){
            holder.interestToBePaid.setVisibility(GONE);
        }else{
            holder.interestToBePaid.setVisibility(View.VISIBLE);
        }
        if(investment.getInvestmentMonth() == 0){
            holder.investmentMonth.setVisibility(GONE);
        }else{
            holder.investmentMonth.setVisibility(View.VISIBLE);
        }
        if(investment.getInvestmentCategory().isEmpty()){
            holder.investmentCategory.setVisibility(GONE);
        }else{
            holder.investmentCategory.setVisibility(View.VISIBLE);
        }
        if(investment.getInvestmentDate()==0){
            holder.investmentDate.setVisibility(GONE);
        }else{
            holder.investmentDate.setVisibility(View.VISIBLE);
        }

        final boolean isExpanded = position==mExpandedPosition;
        Log.d("Tracker","Visibility is "+isExpanded + " : "+position);
        holder.otherDetails.setVisibility(isExpanded?View.VISIBLE:View.GONE);
        holder.itemView.setActivated(isExpanded);

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mExpandedPosition = isExpanded ? -1:position;

                if(investment.getInvestmentPercent()==0.0 && investment.getInvestmentMonth() == 0 && investment.getInvestmentCategory().isEmpty() && investment.getInvestmentDate()==0){
                    Toast.makeText(holder.itemView.getContext(),"No more information to show. You can long press on this investment and add some more details!",Toast.LENGTH_LONG).show();
                }
               notifyDataSetChanged();
            }
        });
    }

    @Override
    public int getItemCount() {
        return InvestmentsList.size();
    }

    public void removeItem(int position) {
        InvestmentsList.remove(position);
        notifyItemRemoved(position);
    }

    public void restoreItem(Investment item, int position) {
        InvestmentsList.add(position, item);
        notifyItemInserted(position);
    }

  /*  *//**
     * Formatting timestamp to `MMM d` format
     * Input: 2018-02-21 00:15:42
     * Output: Feb 21
     *//*
    private String formatDate(String dateStr) {
        try {
            SimpleDateFormat fmt = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss",Locale.ENGLISH);
            Date date = fmt.parse(dateStr);
            SimpleDateFormat fmtOut = new SimpleDateFormat("MMM d",Locale.ENGLISH);
            return fmtOut.format(date);
        } catch (ParseException e) {
            e.printStackTrace();
        }

        return "";
    }*/
}

File: app/src/main/java/com/soumikshah/investmenttracker/view/InvestmentAdapter.kt
package com.soumikshah.investmenttracker.view

import android.content.Context
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.text.HtmlCompat
import androidx.recyclerview.widget.RecyclerView
import com.soumikshah.investmenttracker.R
import com.soumikshah.investmenttracker.database.model.Investment
import java.text.SimpleDateFormat
import java.util.*

class InvestmentAdapter
/**
 * Formatting timestamp to `MMM d` format
 * Input: 2018-02-21 00:15:42
 * Output: Feb 21
 */
/*
      private String formatDate(String dateStr) {
          try {
              SimpleDateFormat fmt = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss",Locale.ENGLISH);
              Date date = fmt.parse(dateStr);
              SimpleDateFormat fmtOut = new SimpleDateFormat("MMM d",Locale.ENGLISH);
              return fmtOut.format(date);
          } catch (ParseException e) {
              e.printStackTrace();
          }
  
          return "";
      }*/(private val context: Context, private val InvestmentsList: ArrayList<Investment>) : RecyclerView.Adapter<InvestmentAdapter.MyViewHolder>() {
    var mExpandedPosition = -1

    class MyViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var investmentName: TextView = view.findViewById(R.id.investment)
        var investmentAmount: TextView = view.findViewById(R.id.investmentAmount)
        var investmentMedium = view.findViewById<TextView>(R.id.investmentMedium)
        var interestToBePaid: TextView = view.findViewById(R.id.interestToBePaid)
        var investmentCategory: TextView = view.findViewById(R.id.investmentCategory)
        var investmentDate: TextView = view.findViewById(R.id.investedDate)
        var investmentMonth: TextView = view.findViewById(R.id.investedNumberOfMonths)
        var viewBackground: RelativeLayout = view.findViewById(R.id.view_background)

        @JvmField
        var viewForeground: RelativeLayout = view.findViewById(R.id.view_foreground)
        var otherDetails: LinearLayout = view.findViewById(R.id.otherDetails)
        var viewBackgroundText: TextView = view.findViewById(R.id.background_text)
        var viewBackgroundImage: ImageView = view.findViewById(R.id.background_image)

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val itemView = LayoutInflater.from(parent.context)
                .inflate(R.layout.investment_list_row, parent, false)
        return MyViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        val investment = InvestmentsList.get(position)
        holder.investmentName.text = investment.investmentName
        holder.investmentAmount.text = String.format(Locale.ENGLISH, "Rs.%,d", investment.investmentAmount)
        val investmentMedium = "Invested in: <b>" + investment.investmentMedium + "</b>"
        val sim = SimpleDateFormat("dd/MM/yyyy", Locale.ENGLISH)
        val investmentCategory = "Investment Type: <b>" + investment.investmentCategory + "</b>"
        val investmentDate = "Invested on: <b>" + sim.format(investment.investmentDate) + "</b>"
        val investmentMonth = "Invested for: <b>" + investment.investmentMonth + "</b>"
        val investmentInterestToBePaid = "Interest to be paid: <b>" + investment.investmentPercent + "</b>"
        holder.investmentMedium.text = HtmlCompat.fromHtml(investmentMedium,HtmlCompat.FROM_HTML_MODE_LEGACY)
        holder.investmentCategory.text = HtmlCompat.fromHtml(investmentCategory,HtmlCompat.FROM_HTML_MODE_LEGACY)
        holder.investmentDate.text = HtmlCompat.fromHtml(investmentDate,HtmlCompat.FROM_HTML_MODE_LEGACY)
        holder.investmentMonth.text = HtmlCompat.fromHtml(investmentMonth,HtmlCompat.FROM_HTML_MODE_LEGACY)
        holder.interestToBePaid.text = HtmlCompat.fromHtml(investmentInterestToBePaid,HtmlCompat.FROM_HTML_MODE_LEGACY)
        if (investment.investmentPercent.toDouble() == 0.0) {
            holder.interestToBePaid.visibility = View.GONE
        } else {
            holder.interestToBePaid.visibility = View.VISIBLE
        }
        if (investment.investmentMonth == 0) {
            holder.investmentMonth.visibility = View.GONE
        } else {
            holder.investmentMonth.visibility = View.VISIBLE
        }
        if (investment.investmentCategory.isEmpty()) {
            holder.investmentCategory.visibility = View.GONE
        } else {
            holder.investmentCategory.visibility = View.VISIBLE
        }
        if (investment.investmentDate == 0L) {
            holder.investmentDate.visibility = View.GONE
        } else {
            holder.investmentDate.visibility = View.VISIBLE
        }
        val isExpanded = position == mExpandedPosition
        holder.otherDetails.visibility = if (isExpanded) View.VISIBLE else View.GONE
        holder.itemView.isActivated = isExpanded
        holder.itemView.setOnClickListener {
            mExpandedPosition = if (isExpanded) -1 else position
            if (investment.investmentPercent.toDouble() == 0.0 && investment.investmentMonth == 0 && investment.investmentCategory.isEmpty() && investment.investmentDate == 0L) {
                Toast.makeText(holder.itemView.context, R.string.toast_message_no_detail, Toast.LENGTH_LONG).show()
            }
            notifyDataSetChanged()
        }
    }

    override fun getItemCount(): Int {
        return InvestmentsList.size
    }

    fun removeItem(position: Int) {
        InvestmentsList.removeAt(position)
        notifyItemRemoved(position)
    }

    fun restoreItem(item: Investment, position: Int) {
        InvestmentsList.add(position, item)
        notifyItemInserted(position)
    } /*  */
}